relations/jf/src/note.rs

```diff
@@ -0,0 +1,143 @@
+use ark_ff::Zero;
+use jf_primitives::circuit::rescue::RescueNativeGadget;
+use jf_relation::{Circuit, PlonkCircuit, Variable};
+
+use crate::{
+    shielder_types::{convert_hash, Note, Nullifier, TokenAmount, Trapdoor},
+    CircuitField, PlonkResult, PublicInput, Relation,
+};
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub enum NoteType {
+    Deposit,
+    Spend,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct NoteRelation {
+    pub note: Note,
+    pub nullifier: Nullifier,
+    pub token_id_var: Variable,
+    pub token_amount: TokenAmount,
```

Comment:
personally I would avoid mixing `Variable` with frontend types - I think it really doesn't hurt to double-create token id variable: especially that we'd gain consistency

Reply:
an alternative would be to allow for both options, i.e. to register either a `Variable` or `TokenId`; in the old code we had a general, type-safe builder-pattern: https://github.com/Cardinal-Cryptography/aleph-node/blob/main/relations/src/shielder/note_var.rs, but this is an overkill I think...

Reply:
changed

Reply:
outdated